### PR TITLE
Persist the scan first then install agents when running a scan using the API

### PR DESCRIPTION
**Pull request by elyousfi5**
Run scan:
[Screencast from 2024-07-02 08-55-33.webm](https://github.com/Ostorlab/oxo/assets/144013278/0e31541d-790e-40b7-974c-c3a66bfdd026)

oxo scan list:
Screenshot from 2024-07-03 10-56-35
Screenshot from 2024-07-03 10-56-16

### CODE REVIEW

**Review by PiranhaSa — APPROVED** [review]
lgtm

File: src/ostorlab/runtimes/local/runtime.py [hunk 1]
```diff
@@ -170,6 +170,23 @@ def _docker_checks(self):
 
         self._docker_client = docker.from_env()
 
+    def prepare_scan(
```
Comment (3asm) [inline src/ostorlab/runtimes/local/runtime.py:173]: As discussed, the API should return once the scan is persisted the scan runtime should start in a dedicated thread or process.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: src/ostorlab/runtimes/local/runtime.py [hunk 2]
```diff
@@ -170,6 +170,23 @@ def _docker_checks(self):
...
+    ) -> None:
+        """Prepare scan entry in the database.
+
+        Args:
+            title: Scan title.
+            assets: The target asset to scan.
+        """
+
+        console.info("Creating scan entry")
+        if assets is None:
```
Comment (benyissa) [inline src/ostorlab/runtimes/local/runtime.py:184]: I would add a unit test for this case.

**Review by benyissa — COMMENTED** [review]
LGTM

File: src/ostorlab/runtimes/local/runtime.py (hunk 2 above)
Comment (elyousfi5) [reply src/ostorlab/runtimes/local/runtime.py:184]: which case man, this is a copy paste (refactoring)

File: tests/serve_app/oxo_test.py [hunk 3]
```diff
@@ -2360,14 +2363,15 @@ def testRunScanMutation_whenNetworkAsset_shouldRunScan(
...
-    assert args["assets"][1].mask == "24"
+    assert prepare_scan_mock.called is True
+    title = prepare_scan_mock.call_args[1]["title"]
+    assets = prepare_scan_mock.call_args[1]["assets"]
+    assert title == "Test Scan Network Asset"
+    assert len(assets) == 2
+    assert assets[0].host == "8.8.8.8"
+    assert assets[0].mask == "32"
+    assert assets[1].host == "8.8.4.4"
+    assert assets[1].mask == "24"
```
Comment (oussamaessaji) [inline tests/serve_app/oxo_test.py:2374]: Assert on the scan id also

**Comment by 3asm** [comment]
The target is not showing in the video?

**Comment by elyousfi5** [comment]
> The target is not showing in the video?

The video wasnt updated after the fix. Please check again.

File: src/ostorlab/runtimes/local/runtime.py [hunk 4]
```diff
@@ -170,6 +170,24 @@ def _docker_checks(self):
...
+            assets_str = "N/A"
+        else:
+            assets_str = f'{", ".join([str(asset) for asset in assets])}'
```
[the remark this answers is not in the document]
Comment (3asm) [reply src/ostorlab/runtimes/local/runtime.py:187]: Why do we need asset str if we have asset passed to it?

File: src/ostorlab/runtimes/local/runtime.py [hunk 5]
```diff
@@ -190,13 +208,8 @@ def scan(
...
         try:
-            console.info("Creating scan entry")
-            if assets is None:
-                assets_str = "N/A"
-            else:
-                assets_str = f'{", ".join([str(asset) for asset in assets])}'
-                # TODO(mohsinenar): we need to add support for storing multiple assets and rename this to target.
-            self._scan_db = self._create_scan_db(asset=assets_str[:255], title=title)
+            if hasattr(self, "_scan_db") is False or self._scan_db is None:
+                self.prepare_scan(title=title, assets=assets)
```
Comment (3asm) [inline src/ostorlab/runtimes/local/runtime.py:212]: Why do we need to check hasattr, this hint at a bug somewhere.

File: src/ostorlab/serve_app/oxo.py [hunk 6]
```diff
@@ -793,6 +794,60 @@ def mutate(
 
         runtime_instance: runtime.LocalRuntime = runtime.LocalRuntime()
         runtime_instance.follow = []
+        created_scan = runtime_instance.prepare_scan(
+            assets=scan_assets, title=scan.title
+        )
+        RunScanMutation._persist_scan_info(created_scan, scan)
+
+        executor = futures.ThreadPoolExecutor(max_workers=1)
```
Comment (3asm) [inline src/ostorlab/serve_app/oxo.py:802]: ThreadPoolExecutor for a single thread, why not use a single thread? or you can pool executor but assign that in the init and use it for all tasks.

File: src/ostorlab/serve_app/oxo.py [hunk 7]
```diff
@@ -793,6 +794,60 @@ def mutate(
...
+            RunScanMutation._run_scan_background,
+            runtime_instance,
+            agent_group,
+            scan,
+            scan_assets,
+        )
+        return RunScanMutation(scan=created_scan)
+
+    @staticmethod
+    def _persist_scan_info(
```
Comment (3asm) [inline src/ostorlab/serve_app/oxo.py:813]: Why static methods?

**Review by 3asm — CHANGES_REQUESTED** [review]

File: src/ostorlab/serve_app/oxo.py (hunk 7 above)
Comment (elyousfi5) [reply src/ostorlab/serve_app/oxo.py:810]: As discussed, these are specific to the mutation so they should be static

File: src/ostorlab/runtimes/local/runtime.py (hunk 5 above)
Comment (elyousfi5) [reply src/ostorlab/runtimes/local/runtime.py]: the init was mocked in some unit tests, i did address that.

**Review by 3asm — APPROVED** [review]

**Pull request merged**